### PR TITLE
Start Maintenance Daemon for Main DB at the server start.

### PR DESCRIPTION
**Pull request by emelsimsek**
DESCRIPTION: This change starts a maintenance deamon at the time of server start if there is a designated main database.

This is the code flow:

1. User designated an admin database:
   `ALTER SYSTEM SET citus.main_db =  "myadmindb";`

2. When postmaster starts, in _PG_Init, citus calls 
    `InitializeMaintenanceDaemonForMainDb`
  
    This function registers a background worker to run `CitusMaintenanceDaemonMain `with `databaseOid = 0 `

3. `CitusMaintenanceDaemonMain ` takes some special actions when databaseOid is 0:
     - Gets the citus.main_db  value.
     - Connects to the  citus.main_db
     - Now the `MyDatabaseId `is available, creates a hash entry for it.
     - Then follows the same control flow as for a regular db,

### CODE REVIEW

File: src/backend/distributed/utils/maintenanced.c [hunk 1]
```diff
@@ -292,45 +329,74 @@ CitusMaintenanceDaemonMain(Datum main_arg)
...
+		}
+
+		before_shmem_exit(MaintenanceDaemonShmemExit, ObjectIdGetDatum(MyDatabaseId));
+		databaseOid = MyDatabaseId;
+		myDbData->userOid = GetSessionUserId();
+	}
+	else
 	{
-		/*
-		 * Another maintenance daemon is running. This usually happens because
```
Comment (JelteF) [inline src/backend/distributed/utils/maintenanced.c:324]: This handling of an edgecase seems to have been removed by accident. And this same edgecase should be handled for the control database.

File: src/backend/distributed/utils/maintenanced.c [hunk 2]
```diff
@@ -146,6 +147,12 @@ InitializeMaintenanceDaemon(void)
 void
 InitializeMaintenanceDaemonForAdminDB(void)
 {
+	if (strcmp(ControlDbName, "") == 0)
```
Comment (JelteF) [inline src/backend/distributed/utils/maintenanced.c:150]: Are GUCs already read in _PG_init? If not, this would always return true.

Comment (emelsimsek) [reply src/backend/distributed/utils/maintenanced.c:150]: Yes, it appears that they are read since this change is able to start a maintenance deamon in the controldb even if there is no connection.

It is also possible to move this check to CitusMaintenanceDaemonMain, Then a background worker will start but then quit immediately.

File: src/backend/distributed/utils/maintenanced.c (hunk 1 above)
Comment (emelsimsek) [reply src/backend/distributed/utils/maintenanced.c:324]: I had merged those two error cases. But yes we have lost the comments section. So I restored.

In If (databaseOid = 0) block workerPid zeroed out by memset. So I am not adding if it is non-zero here check.

File: src/backend/distributed/shared_library_init.c [hunk 3]
```diff
@@ -481,6 +481,7 @@ _PG_init(void)
 #endif
 
 	InitializeMaintenanceDaemon();
+	InitializeMaintenanceDaemonForMainDb();
```
[the remark this answers is not in the document]
Comment (emelsimsek) [reply src/backend/distributed/shared_library_init.c:484]: Both styles seem to be used in the codebase. E.g:

https://github.com/citusdata/citus/commit/bcf54c50141c54825f88e9243abc5755e3a8ec0a#:~:text=MaintenanceDaemonDBData%20*)-,MaintenanceDaemonDBData%20*myDbData%20%3D%20(MaintenanceDaemonDBData%20*),HASH_FIND%2C%20NULL)%3B,-if%20(!myDbData

I do not have a preference.

File: src/test/regress/citus_tests/test/test_maintenancedeamon.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# This test checks that once citus.main_db is set and the
+# server is restarted. A Citus Maintenance Daemon for the main_db
+# is launched. This should happen even if there is no query run
+# in main_db yet.
+def test_set_maindb(coord):
+    with coord.cur() as cur1:
+        cur1.execute("DROP DATABASE IF EXISTS mymaindb;")
```
Comment (JelteF) [inline src/test/regress/citus_tests/test/test_maintenancedeamon.py:7]: Is this line really necessary?

File: src/test/regress/citus_tests/test/test_maintenancedeamon.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        cur1.execute("CREATE DATABASE mymaindb;")
```
Comment (JelteF) [inline src/test/regress/citus_tests/test/test_maintenancedeamon.py:8]: Let's track the databases we create so we can clean them up automatically, just like we do for schemas. See `create_schema`, `cleanup_schemas` and `cleanup_test_leftovers` in `src/test/regress/citus_tests/common.py`

File: src/backend/distributed/utils/maintenanced.c [hunk 6]
```diff
@@ -139,6 +140,49 @@ InitializeMaintenanceDaemon(void)
...
+		return;
+	}
+
+	BackgroundWorker worker;
+
+	memset(&worker, 0, sizeof(worker));
+
+	SafeSnprintf(worker.bgw_name, sizeof(worker.bgw_name),
+				 "Citus Maintenance Daemon: %u/%u",
+				 0, 0);
```
Comment (JelteF) [inline src/backend/distributed/utils/maintenanced.c:162]: I think instead of 0/0 it's more useful to users to write out what it is for.
```suggestion
	strcpy_s(worker.bgw_name, sizeof(worker.bgw_name),
				"Citus Maintenance Daemon for Main DB");
```

File: src/backend/distributed/utils/maintenanced.c [hunk 7]
```diff
@@ -292,45 +336,101 @@ CitusMaintenanceDaemonMain(Datum main_arg)
...
 
-	if (myDbData->workerPid != 0)
+		if (!myDbData)
+		{
+			/*
+			 * When the database crashes, background workers are restarted, but
+			 * the state in shared memory is lost. In that case, we exit and wait
+			 * for Postmaster calling __PG_Init which in turn calls
+			 * InitializeMaintenanceDaemonForAdminDB.
+			 */
```
[the remark this answers is not in the document]
Comment (JelteF) [reply src/backend/distributed/utils/maintenanced.c:370]: I actually think it might be good to add this insertion logic (including zero-ing) to a function that is called here and by `InitializeMaintenanceDaemonBackend`.

[the remark this answers is not in the document]
Comment (JelteF) [reply src/backend/distributed/utils/maintenanced.c]: I think my main source of finding this code confusing is the two branches (`if (databaseOid == 0)` and `else`) that seem similar, but are actually quite different. I think it would help readability to split these a bit better. Either by putting their bodies in a function, or creating a new `CitusMaintenanceDaemonMain` function for the databaseOid == 0 case. These could then call some functions for the shared code. But I think having these long branches but quite different branches make the code hard to follow, since now there's two paths that you need to look at carefully.

File: src/test/regress/citus_tests/test/test_maintenancedeamon.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    wait_until_maintenance_deamons_start(1, cluster)
+
+    assert (
+        cluster.coordinator.sql_value(
+            "SELECT count(*) FROM pg_stat_activity WHERE application_name = 'Citus Maintenance Daemon' AND datname='mymaindb';"
+        )
+        == 0
+    )
+
+    cluster.coordinator.cleanup_databases()
```
Comment (JelteF) [inline src/test/regress/citus_tests/test/test_maintenancedeamon.py:76]: `cleanup_databases` should be called in the `cleanup_test_leftovers` function so that it is called even when the test fails.

File: src/backend/distributed/utils/maintenanced.c [hunk 9]
```diff
@@ -271,66 +339,97 @@ WarnMaintenanceDaemonNotStarted(void)
...
- * CitusMaintenanceDaemonMain is the maintenance daemon's main routine, it'll
- * be started by the background worker infrastructure.  If it errors out,
- * it'll be restarted after a few seconds.
+ * ConnectToDatabase connects to the database for the given databaseOid.
+ * if databaseOid is 0, connects to MainDb and then creates a hash entry.
+ * If a hash entry cannot be created for MainDb it exists the process requesting a restart.
+ * However for regular databases, it exits without requesting a restart since another
+ * subsequent backend should be starting the Maintenance Deamon.
+ * If a hash entry found has a valid workerPid, it exits
+ * without requesting a restart since there is already a deamon running.
```
Comment (JelteF) [inline src/backend/distributed/utils/maintenanced.c:348]: rewording + typo fixes

```suggestion
 * ConnectToDatabase connects to the database for the given databaseOid.
 * if databaseOid is 0, connects to MainDb and then creates a hash entry.
 * If a hash entry cannot be created for MainDb it exists the process requesting a restart.
 * However for regular databases, it exits without requesting a restart since another
 * subsequent backend should be starting the Maintenance Daemon.
 * If the found hash entry has a valid workerPid, it exits
 * without requesting a restart since there is already a daemon running.
```

**Review by JelteF — APPROVED** [review]
Left a small suggestions about the wording in the codecomment. But other than that this looks good now.
